feat: added Check for embedded CSS and the correct stylesheet in Typescript files

IITI-tushar opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.

-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**

feature

**Issue Number:**#2907

Fixes #2907 

**Snapshots/Videos:**

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**

<!--Add link to Talawa-Docs.-->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->



**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**
yes
<!--Yes or No-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Enhanced GitHub Actions workflow to check CSS imports in modified TypeScript files.
  - Introduced a Python script for automated validation of CSS imports during pull requests.

- **Documentation**
  - Minor update to the documentation comment in the AddOnEntry component.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

IITI-tushar commented:
@palisadoes i have updated the code to run for files changed in the PR...
like here in this workflow the css_check.py didn't ran since there's not changed file...
![image](https://github.com/user-attachments/assets/053d0ad4-1af5-44b9-8bc3-240dcc5fb590)

IITI-tushar commented on .github/workflows/pull-request.yml:
```diff
@@ -40,6 +40,29 @@
+          ./.github/workflows/css_check.py --files ${{ steps.changed-ts-files.outputs.modified_files }} || {
+            echo "Error: CSS check failed"
+            exit 1
+          }
```
done

IITI-tushar commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+    parser = argparse.ArgumentParser(
+        description="Check for CSS violations in TypeScript files."
+    )
+    parser.add_argument("--directory", required=True, help="Directory to check.")
+    parser.add_argument(
+        "--exclude_files",
+        nargs="*",
+        default=[],
+        help="Specific files to exclude from analysis.",
+    )
```
made changes in pull-request.yml... so no need here

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+                        f"Embedded CSS found in {file_path}: {', '.join(embedded_css)}"
+                    )
+
+    return violations, correct_css_imports, embedded_css_violations
```
Return a namedtuple. There is less risk of mistakes later by other contributors

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+        print("\n".join(violations))
+
+    if embedded_css_violations:
+        print("\nEmbedded CSS Violations:")
```
1. Use a single string and at the end give an explanation of what the contributor must do. Use the python """ string representation to achieve this.
    1. We need contributors to understand the next steps and fix things without having to ask why
    2. It reduces the workload of the reviewers
3. Give the color code and line number in the message. Make it easier for the end user to rectify the situation

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+    )
+
+    if violations:
+        print("\nCSS Import Violations:")
```
1. Use a single string and at the end give an explanation of what the contributor must do. Use the python """ string representation to achieve this.
    1. We need contributors to understand the next steps and fix things without having to ask why
    2. It reduces the workload of the reviewers

IITI-tushar replied:
done

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+
+    if correct_css_imports:
+        print("\nCorrect CSS Imports:")
+        print("\n".join(correct_css_imports))
```
Make printing successes an optional message based on a argparse CLI flag.

IITI-tushar replied:
done

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+        print("\nCorrect CSS Imports:")
+        print("\n".join(correct_css_imports))
+    else:
+        print("\nNo correct CSS imports found.")
```
This isn't necessary. Please remove

IITI-tushar replied:
done

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+    if violations or embedded_css_violations:
+        sys.exit(1)  # Exit with error code if violations found
+    else:
+        print("\nNo CSS violations found.")
```
Reporting success isn't necessary. Please remove.

IITI-tushar replied:
done

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+                embedded_css = check_embedded_css(content)
+                if embedded_css:
+                    embedded_css_violations.append(
+                        f"Embedded CSS found in {file_path}: {', '.join(embedded_css)}"
```
Make this a named tuple with only the variable names. Add the text at the time of printing.

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+                        )
+                    else:
+                        violations.append(
+                            f"Invalid CSS import ({css_file}) in {file_path}"
```
Make this a named tuple with only the variable names. Add the text at the time of printing.

IITI-tushar replied:
done

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+                    # Check if the CSS import matches the allowed patterns
+                    if any(css_file.endswith(pattern) for pattern in allowed_css_patterns):
+                        correct_css_imports.append(
+                            f"Correct CSS import ({css_file}) in {file_path}"
```
Make this a named tuple with only the variable names. Add the text at the time of printing.

IITI-tushar replied:
done

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+
+                # Check for CSS imports with an improved regex pattern
+                css_imports = re.findall(
+                    r'import\s+.*?["\'](.*?\.css)["\'];', content
```
There is no test to see whether the CSS file actually exists. Please fix.

palisadoes commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,147 @@
+                    continue
+
+                # Check for CSS imports with an improved regex pattern
+                css_imports = re.findall(
```
How do you handle the use of single or double quotes? What does prettier apply, if relevant?

palisadoes reviewed (changes requested):
See comments

IITI-tushar commented:
![image](https://github.com/user-attachments/assets/fa7ac67d-fb14-453c-8a74-9257a5f58cff)
at last it ran fine....
@palisadoes sir i have addressed your comments have a look

IITI-tushar commented on .github/workflows/css_check.py:
```diff
@@ -0,0 +1,167 @@
+                        # If not found, try to find it relative to the src directory
+                        src_dir = os.path.abspath(directory)
+                        css_file_path = os.path.join(src_dir, css_file)
+
```
done

palisadoes commented:
1. Please make CodeRabbit.ai approve your changes.
2. Fix the failing tests

palisadoes commented:
Move your workflows/scripts/ to the scripts folder

palisadoes commented on .github/workflows/pull-request.yml:
```diff
@@ -40,6 +40,42 @@ jobs:
+
+          echo "${{ steps.changed-ts-files.outputs.modified_files }}" | tr ' ' '\n' | while read -r file; do
+            if [ ! -f "$file" ]; then
+              echo "Error: File not found: $file"
```
All this logic is unnecessary. The workflow will fail in all these scenarios if the conditions aren't correct. Just pass the changed files to the script

IITI-tushar replied:
done and fixed

palisadoes reviewed (changes requested):
See comment

IITI-tushar commented:
@palisadoes 
sir will this help
```
``# Run the CSS import check script on changed .ts and .tsx files
- name: Check for CSS violations and print correct imports
  if: ${{ steps.changed-ts-files.outputs.any_changed == 'true' }}
  run: |
    # Split the modified files into an array
    modified_files=(${ { steps.changed-ts-files.outputs.modified_files } })
    
    # Loop through each modified file
    for file in "${modified_files[@]}"; do
      # Get the directory of the file
      dir=$(dirname "$file")
      # Run the CSS check script with the directory as an argument
      python3 ./.github/workflows/scripts/css_check.py --directory "$dir" --allowed_css_patterns app.module.css
    done`
```

palisadoes commented:
Change  argparse to allow multiple directories and/or files

`nargs`

You don't need to loop

IITI-tushar commented:
@palisadoes Thanks sir, this time it ran in  changed files only....
![image](https://github.com/user-attachments/assets/16f66c73-9db5-42a1-8de2-904bd20d1b6f)

palisadoes commented:
1. Merge your code with the latest upstream.
2. The latest black configuration should be there.
3. Run black outside of your IDE

palisadoes reviewed (changes requested):
The logic must be:

1. Inspect the files provided in the CLI for the embedded CSS and fail if found
1. Fail if the files provided do not have a single import of the validated path to `app.module.css`
1. Fail if a css file is found in the same directory as the submitted file
1. Exclude directories provided in the CLI
1. Only check files that are part of the single root directory provided in the CLI

IITI-tushar commented on .github/workflows/scripts/css_check.py:
```diff
@@ -0,0 +1,205 @@
+                    continue
+
+                # Process TypeScript files
+                if file.endswith((".ts", ".tsx")) and "test" not in root:
```
done

IITI-tushar commented:
![image](https://github.com/user-attachments/assets/cae598b1-9f16-4138-9758-9069f69882e5)
@palisadoes sir how can i resolve this...
` `
   """Main function to run the CSS check.

    This function serves as the entry point to run the CSS check, processing the
    necessary files and directories, and printing the violations found.

    Args:
        None: This function does not take any arguments.
               It uses argparse to handle command-line arguments.

    Returns:
        None: This function does not return any value but prints the violations.
    """`
`
what am i doing wrong with the arguments here...

palisadoes commented:
> ![image](https://private-user-images.githubusercontent.com/165766280/404683808-cae598b1-9f16-4138-9758-9069f69882e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczOTU3ODMsIm5iZiI6MTczNzM5NTQ4MywicGF0aCI6Ii8xNjU3NjYyODAvNDA0NjgzODA4LWNhZTU5OGIxLTlmMTYtNDEzOC05NzU4LTkwNjlmNjk4ODJlNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyMFQxNzUxMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZWFjY2I3NWVjOGJjNDBlZWZlZTI2ODcxOWEwYzM5YWRjZjg4YzAyOTkyMmFhMWVlMTUyNjg4ZGJiODFjMGVmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.o6Cwt9e73UabgtWPUEsgFVJcxlZWCRg0E-i4bzqhpRc) @palisadoes sir how can i resolve this... ` ` """Main function to run the CSS check.
> 
> ```
> This function serves as the entry point to run the CSS check, processing the
> necessary files and directories, and printing the violations found.
> 
> Args:
>     None: This function does not take any arguments.
>            It uses argparse to handle command-line arguments.
> 
> Returns:
>     None: This function does not return any value but prints the violations.
> """`
> ```
> 
> ` what am i doing wrong with the arguments here...

Only have `None`  without the colon or explanations

palisadoes commented:
> logic

You haven't applied this logic. For example, you have a `--directories` flag that accepts files. There are many other suggestions.

palisadoes commented:
You have not applied this as suggested previously

The logic must be:

1. Inspect the files provided in the CLI for the embedded CSS and fail if found
1. Fail if the files provided do not have a single import of the validated path to `app.module.css`
1. Fail if a css file is found in the same directory as the submitted file
1. Exclude directories provided in the CLI
1. Only check files that are part of the single root directory provided in the CLI

You also have a `--directories` CLI option that only takes filenames. This is confusing. Please fix.

palisadoes reviewed (changes requested):
You have not applied this as suggested previously

The logic must be:

1. Inspect the files provided in the CLI for the embedded CSS and fail if found
1. Fail if the files provided do not have a single import of the validated path to `app.module.css`
1. Fail if a css file is found in the same directory as the submitted file
1. Exclude directories provided in the CLI
1. Only check files that are part of the single root directory provided in the CLI

You also have a `--directories` CLI option that only takes filenames. This is confusing. Please fix.

IITI-tushar commented:
@palisadoes should I remove --directories flag and add a --files flag??

palisadoes commented:
> @palisadoes should I remove --directories flag and add a --files flag??

Yes. You should also add a flag stating what the starting point should be to determine whether the CSS import is correct.

IITI-tushar commented on .github/workflows/scripts/css_check.py:
```diff
@@ -0,0 +1,190 @@
+
+    # Normalize exclude paths
+    exclude_files = set(os.path.abspath(file) for file in exclude_files)
+    exclude_directories = set(os.path.abspath(dir) for dir in exclude_directories)
```
outdated

IITI-tushar commented on .github/workflows/scripts/css_check.py:
```diff
@@ -0,0 +1,190 @@
+        exclude_files=args.exclude_files,
+        exclude_directories=args.exclude_directories,
+        allowed_css_patterns=args.allowed_css_patterns,
+    )
```
outdated

IITI-tushar commented on .github/workflows/scripts/css_check.py:
```diff
@@ -0,0 +1,244 @@
+                if file_path in exclude_files:
+                    continue
+
+                if file.endswith((".ts", ".tsx")) and "test" not in root:
```
done

IITI-tushar commented:
@palisadoes sir here's an example of embedded css warning which the script will show...
![image](https://github.com/user-attachments/assets/a21476e3-3c82-42c2-b9f1-eda7577d84b8)
the above warning is for updateSession.tsx

![image](https://github.com/user-attachments/assets/819c6410-7ba1-424a-82dc-4425857cfbea)

is it okay?? 
or do i make some changes in it...

palisadoes commented on .github/workflows/pull-request.yml:
```diff
@@ -40,6 +40,24 @@ jobs:
+            **/*.ts
+            **/*.tsx
+
+      # Run the CSS import check script on changed .ts and .tsx files
```
Move this section below the changed file section below line 64.

IITI-tushar replied:
done

palisadoes commented on .github/workflows/pull-request.yml:
```diff
@@ -40,6 +40,24 @@ jobs:
           chmod +x ./.github/workflows/scripts/countline.py
           ./.github/workflows/scripts/countline.py --lines 600 --exclude_files src/screens/LoginPage/LoginPage.tsx src/GraphQl/Queries/Queries.ts src/screens/OrgList/OrgList.tsx src/GraphQl/Mutations/mutations.ts src/components/EventListCard/EventListCardModals.tsx src/components/TagActions/TagActionsMocks.ts src/utils/interfaces.ts src/screens/MemberDetail/MemberDetail.tsx
 
+      # Get Changed .ts and .tsx Files
```
Delete this. It's duplicated on line 61

IITI-tushar replied:
done

palisadoes reviewed (changes requested):
See comments

palisadoes commented:
The script just needs to print the file name, the color code and the line on which the embedded code was found. One instance per line. No multi-line errors per violation, it makes it harder to read.

```
Flie X, has embedded color code Y on line #. Use a CSS variable in src/style/app.module.css
```

palisadoes commented:
Please make coderabbit.ai approve your changes

IITI-tushar commented on .github/workflows/pull-request.yml:
```diff
@@ -43,6 +43,16 @@
         uses: tj-actions/changed-files@v45
+        
+        # Run the CSS import check script on changed .ts and .tsx files
+      - name: Check for CSS violations and print correct imports
+        if: ${{ steps.changed-files.outputs.any_changed == 'true' }}
+        run: |
+          # Pass all modified files as arguments to the script
+          python3 ./.github/workflows/scripts/css_check.py \
+            --files ${{ steps.changed-files.outputs.modified_files }} \
+            --allowed_css_patterns app.module.css
```
@palisadoes should i add these changes or not??

IITI-tushar commented:
> The script just needs to print the file name, the color code and the line on which the embedded code was found. One instance per line. No multi-line errors per violation, it makes it harder to read.
> 
> ```
> Flie X, has embedded color code Y on line #. Use a CSS variable in src/style/app.module.css
> ```

i have fixed this... the embedded css violations will look like this now..
![image](https://github.com/user-attachments/assets/5e3162bd-c3f2-449a-9bfe-8d2791d04b7d)

palisadoes commented:
> > The script just needs to print the file name, the color code and the line on which the embedded code was found. One instance per line. No multi-line errors per violation, it makes it harder to read.
> > ```
> > Flie X, has embedded color code Y on line #. Use a CSS variable in src/style/app.module.css
> > ```
> 
> i have fixed this... the embedded css violations will look like this now.. ![image](https://private-user-images.githubusercontent.com/165766280/406663744-5e3162bd-c3f2-449a-9bfe-8d2791d04b7d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc4MTE1MTEsIm5iZiI6MTczNzgxMTIxMSwicGF0aCI6Ii8xNjU3NjYyODAvNDA2NjYzNzQ0LTVlMzE2MmJkLWMzZjItNDQ5YS05YmZlLThkMjc5MWQwNGI3ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyNVQxMzIwMTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YjA2MDkyMzYzOGU0MmIyOWE1OTQyNWQ5NWVjNWM4NmI1ZTdmODc2YWYwZmM3ZDJlMmJhMjk1NzE4NGNkZmExJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.AX63hjjRZMzppNnuHO0LZnKsL8wxw-lSk36Ub0t4fhY)

1. The filename must start relative to the root of the repository, not the full file path
2. The error message doesn't state "line # x", you just have a number. The end user won't understand
3. You haven't said "color code" in the message. The end user won't understand.

The less questions reviewers have to answer, the better

palisadoes commented:
1. Make sure code rabbit approves the PR
2. Please provide the latest error message screen shots

IITI-tushar commented:
@palisadoes  sir is it okay..?
```
`Embedded CSS Violations:
- File: src\index.tsx, Line: 41: Embedded color code `#31bb6b` detected. Please replace it with a CSS variable in `src/style/app.module.css`.
- File: src\components\EventManagement\EventAttendance\EventStatistics.tsx, Line: 186: Embedded color code `#008000` detected. Please replace it with a CSS variable in `src/style/app.module.css`.
- File: src\components\EventManagement\EventAttendance\EventStatistics.tsx, Line: 192: Embedded color code `#008000` detected. Please replace it with a CSS variable in `src/style/app.module.css`.`
```

palisadoes commented:
Yes

IITI-tushar commented:
@palisadoes sir i still have to commit one thing in css_check.py.
Can I make a new PR for that?

palisadoes commented:
Yes
